influxql/functions.go

```diff
@@ -1035,6 +1082,412 @@ func ReduceLast(values []interface{}) interface{} {
+		if k1 != k2 {
+			return k1 < k2
+		}
+		return t.lessKey(i, j)
+	}
+	return t.less(i, j, sortFloat, sortInt64, sortUint64)
+}
+
+type topReduceOut struct {
+	positionOut
+}
+
+func (t topReduceOut) Len() int      { return len(t.points) }
+func (t topReduceOut) Swap(i, j int) { t.points[i], t.points[j] = t.points[j], t.points[i] }
+func (t topReduceOut) Less(i, j int) bool {
+	// Now sort by time first, not value
+	sortFloat := func(d1, d2 float64) bool {
+		k1, k2 := t.points[i].Time, t.points[j].Time
+		if k1 != k2 {
+			return k1 < k2
+		}
+		if d1 != d2 {
+			return d1 > d2
+		}
+		return t.lessKey(i, j)
+	}
+
+	sortInt64 := func(d1, d2 int64) bool {
+		k1, k2 := t.points[i].Time, t.points[j].Time
+		if k1 != k2 {
+			return k1 < k2
+		}
+		if d1 != d2 {
+			return d1 > d2
+		}
+		return t.lessKey(i, j)
+	}
+
+	sortUint64 := func(d1, d2 uint64) bool {
+		k1, k2 := t.points[i].Time, t.points[j].Time
+		if k1 != k2 {
+			return k1 < k2
+		}
+		if d1 != d2 {
+			return d1 > d2
+		}
+		return t.lessKey(i, j)
+	}
+	return t.less(i, j, sortFloat, sortInt64, sortUint64)
+}
+
+// callArgs will get any additional field/tag names that may be needed to sort with
+// it is important to maintain the order of these that they were asked for in the call
+// for sorting purposes
+func topCallArgs(c *Call) []string {
+	var names []string
+	for _, v := range c.Args[1 : len(c.Args)-1] {
+		if f, ok := v.(*VarRef); ok {
+			names = append(names, f.Val)
+		}
+	}
+	return names
+}
+
+// MapTop emits the top data points for each group by interval
+func MapTop(itr Iterator, c *Call) interface{} {
+	// Capture the limit if it was specified in the call
+	lit, _ := c.Args[len(c.Args)-1].(*NumberLiteral)
+	limit := int64(lit.Val)
+
+	// Simple case where only value and limit are specified.
+	if len(c.Args) == 2 {
+		out := positionOut{callArgs: topCallArgs(c)}
+
+		for k, v := itr.Next(); k != -1; k, v = itr.Next() {
+			t := k
+			if bt := itr.TMin(); bt > -1 {
+				t = bt
+			}
+			out.points = append(out.points, positionPoint{t, v, itr.Tags()})
+		}
+
+		// If we have more than we asked for, only send back the top values
+		if int64(len(out.points)) > limit {
+			sort.Sort(topMapOut{out})
+			out.points = out.points[:limit]
+		}
+		if len(out.points) > 0 {
+			return out.points
+		}
+		return nil
+	}
+	// They specified tags in the call to get unique sets, so we need to map them as we accumulate them
+	outMap := make(map[string]positionOut)
+
+	mapKey := func(args []string, fields map[string]interface{}, keys map[string]string) string {
+		key := ""
+		for _, a := range args {
+			if v, ok := fields[a]; ok {
+				key += a + ":" + fmt.Sprintf("%v", v) + ","
+				continue
+			}
+			if v, ok := keys[a]; ok {
+				key += a + ":" + v + ","
+				continue
+			}
+		}
+		return key
+	}
+
+	for k, v := itr.Next(); k != -1; k, v = itr.Next() {
+		t := k
+		if bt := itr.TMin(); bt > -1 {
+			t = bt
+		}
+		callArgs := c.Fields()
+		tags := itr.Tags()
+		// TODO send fields in
```

Comment:
I don't know what the TODO actually means.  Maybe elaborate a bit more on what is missing and needs to be done?